pkg/frontend/querymiddleware/roundtrip.go

```diff
@@ -288,7 +293,13 @@ func newQueryTripperware(
 		// It means that the first roundtrippers defined in this function will be the last to be
 		// executed.
 
-		queryHandler := NewHTTPQueryRequestRoundTripperHandler(next, codec, log)
+		var queryHandler MetricsQueryHandler
+
+		if useRemoteExecution {
+			queryHandler = NewEngineQueryRequestRoundTripperHandler(streamingEngine, log)
```

Comment:
~I don't quite follow this part. Even if remote execution is enabled, we still can't use it for queries that have been sharded. Don't we need to fallback to an existing roundtripper in that case?~

Nevermind, IIUC this will never be used when a query is sharded since the sharding middleware will use its own roundtripper.

Reply:
This will still be used if the query is sharded. If the query is sharded, the individual sharded legs will be executed by this roundtripper.